src/wallet/walletutil.cpp

```diff
@@ -57,23 +57,48 @@ static bool IsBerkeleyBtree(const fs::path& path)
 std::vector<fs::path> ListWalletDir()
 {
     const fs::path wallet_dir = GetWalletDir();
+    const fs::path data_dir = GetDataDir();
+    const fs::path blocks_dir = GetBlocksDir();
+
     const size_t offset = wallet_dir.string().size() + 1;
     std::vector<fs::path> paths;
     boost::system::error_code ec;
 
+    bool skip_path;
+    // Here we place the top level dirs we want to skip in case walletdir is datadir or blocksdir
+    const std::vector<fs::path> ignore_paths = {
+                                        blocks_dir,
+                                        data_dir / "testnet3",
+                                        data_dir / "regtest",
+                                        data_dir / "chainstate",
+                                        data_dir / "database",
+                                        data_dir / "indexes",
+                                        };
+
     for (auto it = fs::recursive_directory_iterator(wallet_dir, ec); it != fs::recursive_directory_iterator(); it.increment(ec)) {
         if (ec) {
             LogPrintf("%s: %s %s\n", __func__, ec.message(), it->path().string());
             continue;
         }
-
-        // Get wallet path relative to walletdir by removing walletdir from the wallet path.
-        // This can be replaced by boost::filesystem::lexically_relative once boost is bumped to 1.60.
-        const fs::path path = it->path().string().substr(offset);
+        // We dont want to iterate trough those special node dirs
+        if (it->status().type() == fs::directory_file) {
+            skip_path = false;
+            for(const auto& fpath: ignore_paths) {
+                if (!skip_path && it->path().string().find(fpath.string()) == 0) {
```

Comment:
Since we're using exact paths in `ignore_paths`, can't we use an exact match? (And then `ignore_paths` can become an `unordered_set` and just use `.count` here...)